File: src/shapes/rapier_capsule_shape_2d.cpp
#include "rapier_capsule_shape_2d.h"

rapier2d::Handle RapierCapsuleShape2D::create_rapier_shape() const {
	return rapier2d::shape_create_capsule((height / 2.0) - radius, radius);
}

void RapierCapsuleShape2D::set_data(const Variant &p_data) {
	ERR_FAIL_COND(p_data.get_type() != Variant::ARRAY && p_data.get_type() != Variant::VECTOR2);

	if (p_data.get_type() == Variant::ARRAY) {
		Array arr = p_data;
		ERR_FAIL_COND(arr.size() != 2);
		height = arr[0];
		radius = arr[1];
	} else {
		Point2 p = p_data;
		radius = p.x;
		height = p.y;
	}

	Point2 he(radius, height * 0.5);
	configure(Rect2(-he, he * 2.0));
}

Variant RapierCapsuleShape2D::get_data() const {
	return Point2(height, radius);
}

real_t RapierCapsuleShape2D::get_moment_of_inertia(real_t p_mass, const Size2 &p_scale) const {
	Vector2 he2 = Vector2(radius * 2.0, height) * p_scale;
	return p_mass * he2.dot(he2) / 12.0;
}

File: src/shapes/rapier_capsule_shape_2d.rs
use godot::classes::physics_server_2d::*;
use godot::prelude::*;

use crate::rapier_wrapper::prelude::*;
use crate::shapes::rapier_shape::*;
use crate::Vector;
pub struct RapierCapsuleShape2D {
    height: f32,
    radius: f32,
    pub base: RapierShapeBase,
}
impl RapierCapsuleShape2D {
    pub fn new(rid: Rid) -> Self {
        Self {
            height: 0.0,
            radius: 0.0,
            base: RapierShapeBase::new(rid),
        }
    }
}
impl IRapierShape for RapierCapsuleShape2D {
    fn get_base(&self) -> &RapierShapeBase {
        &self.base
    }

    fn get_mut_base(&mut self) -> &mut RapierShapeBase {
        &mut self.base
    }

    fn get_type(&self) -> ShapeType {
        ShapeType::CAPSULE
    }

    fn get_moment_of_inertia(&self, mass: f32, scale: Vector) -> f32 {
        let he2 = Vector::new(self.radius * 2.0, self.height) * scale;
        mass * he2.dot(he2) / 12.0
    }

    fn allows_one_way_collision(&self) -> bool {
        true
    }

    fn create_rapier_shape(&mut self) -> Handle {
        shape_create_capsule((self.height / 2.0) - self.radius, self.radius)
    }

    fn set_data(&mut self, data: Variant) {
        match data.get_type() {
            VariantType::ARRAY => {
                let arr: Array<f32> = data.to();
                if arr.len() != 2 {
                    return;
                }
                self.height = arr.at(0);
                self.radius = arr.at(1);
            }
            VariantType::VECTOR2 => {
                let vector_data: Vector2 = data.to();
                self.height = vector_data.y;
                self.radius = vector_data.x;
            }
            _ => {
                godot_error!("Invalid shape data");
                return;
            }
        }
        let he = Vector2::new(self.radius, self.height * 0.5);
        let rect = Rect2::new(-he, he * 2.0);
        let handle = self.create_rapier_shape();
        self.base.set_handle(handle, rect);
    }

    fn get_data(&self) -> Variant {
        Vector2::new(self.radius, self.height).to_variant()
    }

    fn get_handle(&mut self) -> Handle {
        self.base.get_handle()
    }
}
